stackpath/structure_compute_network_allocation_claim.go

```diff
@@ -0,0 +1,236 @@
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// convert from the terraform data structure to the allocation data structure we need for API calls
+func convertComputeNetworkAllocationClaim(data *schema.ResourceData) *ipam_models.V1AllocationClaim {
+	return &ipam_models.V1AllocationClaim{
+		Name: data.Get("name").(string),
+		Slug: data.Get("slug").(string),
+		Metadata: &ipam_models.Metav1Metadata{
+			Annotations: convertToMetaV1StringMap(data.Get("annotations").(map[string]interface{})),
+			Labels:      convertToMetaV1StringMap(data.Get("labels").(map[string]interface{})),
+			Version:     data.Get("version").(string),
+		},
+		Spec: &ipam_models.V1AllocationClaimSpec{
+			PrefixLength:    int32(data.Get("prefix_length").(int)),
+			IPFamily:        convertComputeNetworkAllocationIPFamily(data.Get("ip_family")),
+			ReclaimPolicy:   convertComputeNetworkAllocationReclaimPolicy(data.Get("reclaim_policy")),
+			ResourceBinding: convertComputeNetworkAllocationResourceBinding(data.Get("resource_binding")),
+			Allocation:      convertComputeNetworkAllocationClaimSpecAllocation(data.Get("allocation")),
+		},
+	}
+}
+
+// convert from the terraform data structure to the allocation claim data structure we need for update API
+func convertComputeNetworkAllocationClaimUpdate(data *schema.ResourceData) *ipam_models.V1AllocationClaim {
+	// prepare data structure with only fields which are allowed to update, passing in
+	// any additional fields which are not allowed to update causes update api to throw
+	// validation error. hence to allow succesful updates for data changes we are preparing
+	// request body only with fields which are allowed to pass in.
+	// We allow updates to:
+	//  - metadata.annotations
+	//  - metadata.labels
+	//  - name
+	//  - spec.reclaimPolicy.action
+	//  - spec.reclaimPolicy.idleRetentionPeriod
+	//  - spec.resourceBinding
+
+	return &ipam_models.V1AllocationClaim{
+		Name: data.Get("name").(string),
+		Metadata: &ipam_models.Metav1Metadata{
+			Annotations: convertToMetaV1StringMap(data.Get("annotations").(map[string]interface{})),
+			Labels:      convertToMetaV1StringMap(data.Get("labels").(map[string]interface{})),
+		},
+		Spec: &ipam_models.V1AllocationClaimSpec{
+			ReclaimPolicy:   convertComputeNetworkAllocationReclaimPolicy(data.Get("reclaim_policy")),
+			ResourceBinding: convertComputeNetworkAllocationResourceBinding(data.Get("resource_binding")),
+		},
+	}
+}
+
+func convertComputeNetworkAllocationResourceBinding(p interface{}) *ipam_models.V1TypedResourceReference {
+	if len(p.([]interface{})) == 0 {
+		return nil
+	}
+
+	resourceBinding := &ipam_models.V1TypedResourceReference{}
+	resourceBindingData := p.([]interface{})[0].(map[string]interface{})
+	if resourceBindingData["type"].(string) != "" {
+		resourceBinding.Type = resourceBindingData["type"].(string)
+	}
+
+	if resourceBindingData["name"].(string) != "" {
+		resourceBinding.Name = resourceBindingData["name"].(string)
+	}
```

Comment:
Should this check if the type resolves before assigning?
```suggestion
        if resourceBindingDataType, ok := resourceBindingData["type"].(string); ok && resourceBindingDataType != "" {
	if resourceBindingData["type"].(string) != "" {
		resourceBinding.Type = resourceBindingDataType
	}

	if resourceBindingDataName, ok := resourceBindingData["name"].(string); ok && resourceBindingDataName != "" {
		resourceBinding.Name = resourceBindingDataName
	}
```

Reply:
I dont think it is required as terraform schema already defines it as string and schema validator performs validation accordingly. it is expected to be string type interface only when it is received here.

Reply:
ok 👍